### ISSUE TITLE
MPI load balancing

### ISSUE THREAD

**Issue by dhubber** [gandalfcode/gandalf#24]
Need to re-write subroutines to walk the domain trees and compute the new load balanced domains

**Comment by dhubber** [comment]
Load balancing works in principle, but still some bug remains (in CodeTiming perhaps?) so deactivated for now

**Comment by giovanni-rosotti** [comment]
No the code timing one is a separate bug...

**Comment by giovanni-rosotti** [comment]
See #45

**Comment by giovanni-rosotti** [comment]
Load balancing is now implemented. I'll leave the issue open though to remember ourselves that we should time how long it takes. If we find it's an important fraction, we might want to switch to something that converges faster than bisection

**Comment by giovanni-rosotti** [comment]
As we found out in the paper, the load balancing is actually the limit to the scaling. This needs a careful assesment.

**Comment by rbooth200** [comment]
The 'best' solution is likely problem dependent too, for example azimuthal domain decomposition might work well for discs. So I agree that this needs a thorough exploration...

**Comment by giovanni-rosotti** [comment]
I agree. But I think some very important would be to force the load balancing to happen only on relatively high timesteps (not necessarily resync, but anyway when a large fraction of particles are active).
Of course this depends on whether the CPU time is dominated by the lowest or the highest levels...

**Comment by rbooth200** [comment]
I agree. It might also be worth separating load-balancing from particle transfer too.
